### PR TITLE
feat: profile page

### LINKED ISSUE #117: Implement "Profile Page"
## Description
Implement the Profile Page.

## Acceptance Criteria

- [ ] Implement the interface as specified in [#92](https://github.com/ipfs-shipyard/pm-idm/issues/92)

- [ ] Implement a 404 page in case the identity does not exist in the wallet

### PR DESCRIPTION
**Pull request by dominguesgm**
First implementation of the profile page. All data is hardcoded, due to missing integration.

This closes https://github.com/ipfs-shipyard/pm-idm/issues/117, unless this issue also considers integration.

### CODE REVIEW

File: src/pages/identity/activity-boxes/Devices.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connectIdmWallet } from 'react-idm-wallet';
+import { DesktopIcon, LaptopIcon, TabletIcon, MobileIcon } from '@nomios/Web-uikit';
+import StatusIndicator from './shared/StatusIndicator';
+import GenericItem from './shared/GenericItem';
+import GenericList from './shared/GenericList';
+
+function iconFromDevicetype(type) {
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/Devices.js]: We have a different approach but with the same purpose on [this file](https://github.com/ipfs-shipyard/nomios-web/blob/master/src/modals/new-identity-flow/common-steps/setup-device-step/devices.js).

Can we use the same approach for both cases? We can use either your or mine approach. But I think we must get this standardized.

File: craco.config.js [hunk 2]
```diff
@@ -126,12 +126,12 @@ module.exports = {
                             },
                             // Uniquify classnames and ids so that they are unique and
                             // don't conflict with each other
-                            {
+                            /* {
```
Comment (PedroMiguelSS) [inline craco.config.js]: Why did you remove this loader?

File: src/pages/identity/IdentityActivity.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            </div>
+        );
+    }
+
+    handleMyDevices = () => {
+        alert('See my devices');
+    };
+
+    handleMySocialProofs = () => {
+        alert('See my devices');
```
Comment (PedroMiguelSS) [inline src/pages/identity/IdentityActivity.js]: Change it this alert to something different, like: `alert('See my social proofs')`

File: src/pages/identity/IdentityActivity.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    };
+
+    handleMyApps = () => {
+        alert('See my devices');
```
Comment (PedroMiguelSS) [inline src/pages/identity/IdentityActivity.js]: Same here, like: `alert('See my apps')`

File: src/pages/identity/IdentityDetails.css [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            margin-left: 5rem;
+
+            & .buttonIcon {
+                font-size: 1.6rem;
+            }
+        }
+    }
+}
+
+@media (max-width: 1280px) {
```
Comment (PedroMiguelSS) [inline src/pages/identity/IdentityDetails.css]: Shouldn't we be using `custom-medias` here? Maybe we should add them to [nomios-web-uikit](https://github.com/ipfs-shipyard/nomios-web-uikit) and import them here.

//cc @satazor

File: src/pages/identity/activity-boxes/Apps.css [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        width: 25rem;
+        margin-top: 0;
+        align-self: flex-start;
+        color: var(--color-armadillo);
+    }
+
+    & .globeSvg {
+        height: 15rem;
+        align-self: flex-end;
+
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/Apps.css]: Can you remove this extra line?

File: src/pages/identity/activity-boxes/Apps.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import GenericItem from './shared/GenericItem';
+import GenericList from './shared/GenericList';
+import { Svg } from '@nomios/Web-uikit';
+import styles from './Apps.css';
+import appIcon from '../../../shared/media/backgrounds/background-pattern-1440p.png';
+
+const globeSvg = import('../../../shared/media/illustrations/globe.svg');
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/Apps.js]: Give the same webpack chunk name as we are doing to other illustrations: `svg-illustrations-sprite`

File: src/pages/identity/activity-boxes/Apps.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            )}
+        </GenericList>
+    );
+};
+
+AppsBox.propTypes = {
+    apps: PropTypes.array,
+};
+
+AppsBox.defaultProps = {
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/Apps.js]: Just to make sure: this will be removed as soon as `idm-wallet` has support for apps, right?

File: src/pages/identity/activity-boxes/Devices.css [hunk 9]
```diff
@@ -0,0 +1,6 @@
+
+@media (max-width: 1280px) {
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/Devices.css]: Same here about `custom-medias`.

File: src/pages/identity/activity-boxes/SocialProofs.css [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+& .separator {
+    width: 100%;
+    height: 0.1rem;
+    opacity: 0.15;
+    background-color: var(--color-armadillo);
+}
+
+@media (max-width: 1280px) {
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/SocialProofs.css]: Same here.

File: src/pages/identity/activity-boxes/SocialProofs.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+import { FacebookIcon, TwitterIcon, LinkedinIcon, GithubIcon, TextButton } from '@nomios/web-uikit';
+import styles from './SocialProofs.css';
+
+class SocialProofItem extends Component {
+    render() {
+        const { network, account } = this.props;
+
+        let socialIcon = null;
+
+        switch (network) {
```
Comment (PedroMiguelSS) [inline src/pages/identity/activity-boxes/SocialProofs.js]: I would change this to a new function which returns the icon.

Like so, your render method would be clearer and it would be like this: 

```js
const socialIcon = this.inferSocialIcon(network);
```

But the change it's up to you, it's a personal preference.

**Review by PedroMiguelSS — CHANGES_REQUESTED** [review]

File: src/pages/identity/activity-boxes/Devices.js (hunk 1 above)
Comment (dominguesgm) [reply src/pages/identity/activity-boxes/Devices.js]: I think it would be a good idea to have it standardized. I'll change it to get the devices from the `devices.js` file

File: craco.config.js (hunk 2 above)
Comment (dominguesgm) [reply craco.config.js]: This loader is being removed because I'm using global css classes to style one of the inner paths of an svg illustration (the `globe`). This loader would generate unique classes which I was unable to use from a css file.

A possible approach that would allow us to mantain this loader would be to access the path to style by using sibling selectors, but I personally prefer the current approach.

@satazor any thoughts on this?

File: src/pages/identity/IdentityDetails.css (hunk 5 above)
Comment (dominguesgm) [reply src/pages/identity/IdentityDetails.css]: I was honestly unfamiliar with `custom-medias`, so I don't really have an opinion here. Wouldn't mind to hear another piece of feedback, in case of consensus I can change it.

File: src/pages/identity/activity-boxes/Apps.js (hunk 8 above)
Comment (dominguesgm) [reply src/pages/identity/activity-boxes/Apps.js]: Yes, this is just an hardcoded example to test the styles 😄

File: src/pages/identity/activity-boxes/SocialProofs.js (hunk 11 above)
Comment (dominguesgm) [reply src/pages/identity/activity-boxes/SocialProofs.js]: No, I agree. Just changed it 😄

File: craco.config.js (hunk 2 above)
Comment (satazor) [reply craco.config.js]: Even with this loader, you should be able to target them with selector prefixes. If that doesn’t work, you may always change the svg path to include attributes, such as data-xxx attributes so that you are able to target them.

File: src/pages/identity/Identity.js [hunk 12]
```diff
@@ -1,10 +1,47 @@
...
+}
 
 Identity.propTypes = {
     match: PropTypes.object.isRequired,
+    isLoaded: PropTypes.func.isRequired,
 };
 
-export default Identity;
+export default connectIdmWallet((idmWallet) => {
+    const isLoaded = () => idmWallet.identities.isLoaded();
```
Comment (satazor) [inline src/pages/identity/Identity.js]: We can completely remove this. My PR at #16 ensures that the app is only mounted if identities are loaded.

File: src/pages/identity/IdentityDetails.css (hunk 5 above)
Comment (satazor) [reply src/pages/identity/IdentityDetails.css]: If we are going to use this media query often, then we should create a custom media. Please ask Sousa if this is something just for this page or not.

File: src/pages/identity/IdentityDetails.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                    </Button>
+                </div>
+            </div>
+        );
+    }
+
+    buildDetailsObject() {
+        const { getProfileDetails, matchId } = this.props;
+
+        const details = getProfileDetails(matchId);
```
Comment (satazor) [inline src/pages/identity/IdentityDetails.js]: Feels strange to mutate an object that is being given to you. Perhaps wrap it or shallow clone it?

File: src/pages/identity/activity-boxes/Devices.js (hunk 1 above)
Comment (satazor) [reply src/pages/identity/activity-boxes/Devices.js]: Be careful in the way you do this. It feels strange for this page to depend on a file of the new identity flow. On the other hand, also be careful to moving that file to the shared folder as it also contains things specific for the new identity flow.

File: src/shell/Shell.css [hunk 14]
```diff
@@ -8,7 +8,8 @@
     }
 
     & .content {
-        flex: 1 0 auto;
-        background: yellow;
+        width: calc(100% - 8rem);
```
Comment (satazor) [inline src/shell/Shell.css]: I think all the positioning of the sidebar and content should be here, including the position fixed. Also, we no longer need the display flex on the shell and flex 0 0 auto on the sidebar.

**Review by satazor — CHANGES_REQUESTED** [review]

File: src/pages/identity/Identity.js [hunk 15]
```diff
@@ -1,10 +1,47 @@
...
+            return null;
+        }
+
+        return (
+            <div className={ styles.splitWrapper }>
+                <div className={ styles.split }>
+                    <div className={ styles.background } style={ { backgroundImage: `url(${backgroundPatternUrl})` } } />
+                    <PageLayout>
+                        <div className={ styles.top }>
+                            <IdentityDetails matchId={ id } />
```
Comment (satazor) [inline src/pages/identity/Identity.js]: It feels strange for this property to be called `matchId`. Why not `id` or `identityId`? Change this everywhere.

File: src/pages/identity/activity-boxes/Devices.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    );
+};
+
+DevicesBox.propTypes = {
+    matchId: PropTypes.string.isRequired,
+    getDevices: PropTypes.func.isRequired,
+};
+
+export default connectIdmWallet((idmWallet) => {
+    const getDevices = (params) => idmWallet.identities.get(params).devices.list();
```
Comment (satazor) [inline src/pages/identity/activity-boxes/Devices.js]: I think this connect can and show be improved:

```js
export default connectIdmWallet((idmWallet) => (ownProps) => ({
    devices: idmWallet.identities.get(ownProps.matchId).devices.list()
});
```

Please use the same concept everywhere, e.g.: in profile.

**Review by satazor — CHANGES_REQUESTED** [review]

File: src/pages/identity/activity-boxes/Devices.js (hunk 16 above)
Comment (dominguesgm) [reply src/pages/identity/activity-boxes/Devices.js]: Will this approach still trigger a rerender if the devices list changes?

File: src/pages/identity/IdentityDetails.js (hunk 13 above)
Comment (dominguesgm) [reply src/pages/identity/IdentityDetails.js]: I agree. However, given this function is being called inside `render`, maybe I should also turn this into a pure component, no? Otherwise, we'll be creating a new object at each rerender, even if the props don't necessarily change

File: craco.config.js (hunk 2 above)
Comment (dominguesgm) [reply craco.config.js]: The approach chosen was to define the following option: `localIdentName: '[name]__[local]'`

File: src/app/shell/sidebar/Sidebar.css [hunk 17]
```diff
@@ -9,8 +9,12 @@
 }
 
 .sidebar {
-    position: relative;
+    position: fixed;
```
Comment (satazor) [inline src/app/shell/sidebar/Sidebar.css]: As I said earlier, I think the position of the sidebar should be defined in the Shell.css file. It's the role of the Shell to position the content and the sidebar correctly, including their z-index.

You may leave the `position: relative` here because there's elements inside the sidebar that are positioned relative to it, but the `position: fixed` defined in the shell will win.

File: src/pages/identity/Identity.js [hunk 18]
```diff
@@ -1,10 +1,47 @@
...
 Identity.propTypes = {
     match: PropTypes.object.isRequired,
+    hasIdentity: PropTypes.func.isRequired,
 };
 
-export default Identity;
+export default connectIdmWallet((idmWallet) => {
+    const hasIdentity = (id) => idmWallet.identities.has(id);
+
+    return () => ({
```
Comment (satazor) [inline src/pages/identity/Identity.js]: Similar to my suggestion earlier today, we may simplify this:

```
export default connectIdmWallet((idmWallet) => (ownProps) => ({
    hasIdentity: idmWallet.identities.has(ownProps.match.params.id);
});
```

File: src/pages/identity/activity-boxes/Apps.js [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+            )}
+        </GenericList>
+    );
+};
+
+AppsBox.propTypes = {
+    apps: PropTypes.array,
+};
+
+// AppsBox.defaultProps = {
```
Comment (satazor) [inline src/pages/identity/activity-boxes/Apps.js]: Can we remove this?

File: src/shared/components/page-layout/index.js [hunk 20]
```diff
@@ -0,0 +1,3 @@
+import PageLayout from './PageLayout';
```
Comment (satazor) [inline src/shared/components/page-layout/index.js]: Nifty pick: can we do `export { default } from './PageLayout`? Check in other files of this PR as well.

File: src/shared/data/devices/devices.js [hunk 21]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { MobileIcon, TabletIcon, LaptopIcon, DesktopIcon, CheckmarkIcon } from '@nomios/web-uikit';
+
+export default [
+    {
+        id: 'mobile',
+        label: 'Mobile',
+        badge: <CheckmarkIcon />,
```
Comment (satazor) [inline src/shared/data/devices/devices.js]: As I said earlier, it feels strange to have the `badge` defined here. The rest of the fields is ok.

File: src/shared/data/devices/index.js [hunk 22]
```diff
@@ -0,0 +1,3 @@
+import devices from './devices';
```
Comment (satazor) [inline src/shared/data/devices/index.js]: Nifty pick: can we do `export { default } from './devices'`?

**Review by satazor — CHANGES_REQUESTED** [review]

**Comment by satazor** [comment]
@dominguesgm Can you rebase with master? #10 was merged and you may now have a modal trigger for it in the "Edit profile" button.

File: src/modals/new-identity-flow/common-steps/setup-device-step/SetupDeviceStep.js [hunk 23]
```diff
@@ -7,7 +7,7 @@ import { Button, TypeGroup, TypeOption, TextInput, EditIcon } from '@nomios/web-
 import FadeContainer from '../../../../shared/components/fade-container';
 import { notEmpty } from '../../../../shared/form-validators';
 import { deviceInfo } from './deviceInfo';
-import devices from './devices';
+import devices from '../../../../shared/data/devices';
```
Comment (satazor) [inline src/modals/new-identity-flow/common-steps/setup-device-step/SetupDeviceStep.js:9]: Nifty pick: move this close to the other shared imports (two lines above).

File: src/modals/new-identity-flow/common-steps/setup-device-step/SetupDeviceStep.js [hunk 24]
```diff
@@ -7,7 +7,7 @@ import { Button, TypeGroup, TypeOption, TextInput, EditIcon } from '@nomios/web-
 import FadeContainer from '../../../../shared/components/fade-container';
 import { notEmpty } from '../../../../shared/form-validators';
 import { deviceInfo } from './deviceInfo';
```
Comment (satazor) [inline src/modals/new-identity-flow/common-steps/setup-device-step/SetupDeviceStep.js]: Nifty pick: Can we rename this file to `device-info.js`?

File: src/pages/identity/Identity.js [hunk 25]
```diff
@@ -1,10 +1,47 @@
-import React from 'react';
+import React, { Component } from 'react';
 import PropTypes from 'prop-types';
+import { connectIdmWallet } from 'react-idm-wallet';
+import styles from './Identity.css';
```
Comment (satazor) [inline src/pages/identity/Identity.js]: Can we improve the order of the imports? e.g.: import of styles should be last.. also shared imports should be the first ones amongst the relative ones.

File: src/pages/identity/activity-boxes/Apps.js [hunk 26]
```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import GenericItem from './shared/GenericItem';
+import GenericList from './shared/GenericList';
+import { Svg } from '@nomios/Web-uikit';
```
Comment (satazor) [inline src/pages/identity/activity-boxes/Apps.js]: Same about the order here of imports here.

File: src/pages/identity/activity-boxes/Devices.js [hunk 27]
```diff
@@ -0,0 +1,74 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connectIdmWallet } from 'react-idm-wallet';
+import StatusIndicator from './shared/StatusIndicator';
+import GenericItem from './shared/GenericItem';
+import GenericList from './shared/GenericList';
+import devices from '../../../shared/data/devices';
```
Comment (satazor) [inline src/pages/identity/activity-boxes/Devices.js]: Same about the order of the imports here.

File: src/pages/identity/activity-boxes/SocialProofs.js [hunk 28]
```diff
@@ -0,0 +1,135 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import StatusIndicator from './shared/StatusIndicator';
+import GenericList from './shared/GenericList';
+import { FacebookIcon, TwitterIcon, LinkedinIcon, GithubIcon, TextButton } from '@nomios/web-uikit';
```
Comment (satazor) [inline src/pages/identity/activity-boxes/SocialProofs.js]: Same about the order of the imports here.

**Review by satazor — CHANGES_REQUESTED** [review]

File: src/shared/data/devices/devices.js (hunk 21 above)
Comment (dominguesgm) [reply src/shared/data/devices/devices.js]: To be fair, since this is only used in the device shared step, and the badge icon is the same for every device, we can completely remove it from here and directly apply it in the step code.

Comment (satazor) [reply src/shared/data/devices/devices.js]: Let’s do it

File: src/pages/identity/Identity.js [hunk 29]
```diff
@@ -1,10 +1,47 @@
...
+import IdentityDetails from './IdentityDetails';
+import IdentityActivity from './IdentityActivity';
+import PageLayout from '../../shared/components/page-layout/PageLayout';
+import backgroundPatternUrl from '../../shared/media/backgrounds/background-pattern-2160p.png';
```
Comment (satazor) [inline src/pages/identity/Identity.js]: Can we use a `url()` in the css instead of importing in the JS file?

**Review by satazor — CHANGES_REQUESTED** [review]

File: src/pages/identity/activity-boxes/Apps.js (hunk 19 above)
Comment (dominguesgm) [reply src/pages/identity/activity-boxes/Apps.js]: 👍

**Comment by dominguesgm** [comment]
This should be ready for final review and merge. It is integrated and the `Edit Profile` button now triggers the edit profile modal

File: craco.config.js [hunk 30]
```diff
@@ -127,9 +127,10 @@ module.exports = {
                             // Uniquify classnames and ids so that they are unique and
                             // don't conflict with each other
                             {
-                                loader: require.resolve('svg-css-modules-loader'),
+                                loader: 'svg-css-modules-loader',
```
Comment (satazor) [inline craco.config.js]: Can we keep the `require.resolve`?

**Review by satazor — APPROVED** [review]
